File: r/tests/testthat/test-conformance-resolve.R
# Conformance: the SHARED golden resolve cases (tests/cases_resolve.csv at
# repo level) must pass unmodified. Empty cells mean "omitted -> default per
# contract"; expect_error rows name the expected condition class suffix.

test_that("shared resolve conformance cases pass unmodified", {
  cases <- read.csv(file.path(datalib_test_repo_tests(), "cases_resolve.csv"),
                    colClasses = "character")
  expect_gt(nrow(cases), 0)
  root <- datalib_test_fixture_root()
  expect_true(dir.exists(root))

  nz <- function(x) !is.na(x) && nzchar(x)
  for (i in seq_len(nrow(cases))) {
    cs <- cases[i, ]
    args <- list(country = cs$country, kind = cs$kind, root = root)
    if (nz(cs$year)) args$year <- cs$year
    if (nz(cs$survey)) args$survey <- cs$survey
    if (nz(cs$collection)) args$collection <- cs$collection
    if (nz(cs$master_version)) args$master_version <- cs$master_version
    if (nz(cs$adaptation_version)) args$adaptation_version <- cs$adaptation_version

    if (nz(cs$expect_error)) {
      expect_error(do.call(datalib_resolve, args),
                   class = paste0("datalib_error_", cs$expect_error),
                   info = paste("case", cs$case_id))
    } else {
      res <- do.call(datalib_resolve, args)
      expect_identical(res$vintage_folder, cs$expect_vintage_folder,
                       info = paste("case", cs$case_id))
    }
  }
})

File: python/tests/test_conformance_resolve.py
"""Golden resolve cases (shared tests/cases_resolve.csv) — must pass unmodified.

Empty year/survey/vintage cells mean "omitted -> default per contract";
lowercase inputs must normalize; expect_error rows must raise a typed error.
"""

from __future__ import annotations

import pytest

from conftest import read_cases

import datalib as dl
from datalib.errors import InvalidArgumentError, NotFoundError

CASES = read_cases("cases_resolve.csv")


@pytest.mark.parametrize("case", CASES, ids=[c["case_id"] for c in CASES])
def test_resolve_case(case: dict[str, str], fixture_root) -> None:
    kwargs = dict(
        country=case["country"],
        year=case["year"] or None,
        survey=case["survey"] or None,
        kind=case["kind"] or "adaptation",
        collection=case["collection"] or None,
        master_version=case["master_version"] or None,
        adaptation_version=case["adaptation_version"] or None,
        root=fixture_root,
    )
    if case["expect_error"]:
        assert case["expect_error"] == "input"
        with pytest.raises((InvalidArgumentError, NotFoundError)):
            dl.resolve(**kwargs)
    else:
        rp = dl.resolve(**kwargs)
        assert rp.vintage_dir.name == case["expect_vintage_folder"]
        assert rp.stem == case["expect_vintage_folder"]
        assert rp.vintage_dir.is_dir()
